packages/jsii-diff/lib/classes-ifaces.ts

```diff
@@ -0,0 +1,203 @@
+    }
+  }
+
+  if (original.docs.isSubclassable && !updated.docs.isSubclassable) {
+    context.mismatches.report(original, 'has gone from @subclassable to non-@subclassable');
+  }
+
+  for (const [origMethod, updatedMethod] of memberPairs(original, original.allMethods, updated, context)) {
+    compareMethod(original, origMethod, updatedMethod, context);
+  }
+
+  for (const [origProp, updatedProp] of memberPairs(original, original.allProperties, updated, context)) {
+    compareProperty(original, origProp, updatedProp, context);
+  }
+
+  // You cannot have added abstract members to the class/interface, as they are
+  // an added burden on potential implementors.
+  //
+  // Only for types that are explicitly marked as intended to be subclassed by customers.
+  if (subclassableType(original)) {
+    noNewAbstractMembers(original, updated, context);
+  }
+}
+
+export function compareStruct(original: reflect.InterfaceType, updated: reflect.InterfaceType, context: ComparisonContext) {
+  // We don't compare structs here; they will be evaluated for compatibility
+  // based on input and output positions.
+  //
+  // However, even if we don't do full property matching (we don't know yet if
+  // the types should be supertype or subtype checked), we still have to check
+  // that all members which used to be present are still present. This is because
+  // for all non-TypeScript languages it is not allowed to specify members
+  // which aren't actually present in the type.
+  //
+  // A single run of memberPairs() with nothing else will do this check.
+  Array.from(memberPairs(original, original.allProperties, updated, context));
+}
+
+function noNewAbstractMembers<T extends reflect.ReferenceType>(original: T, updated: T, context: ComparisonContext) {
+    const absMemberNames = new Set(updated.allMembers.filter(m => m.abstract).map(m => m.name));
+    const originalMemberNames = new Set(original.allMembers.map(m => m.name));
+    for (const name of absMemberNames) {
+      if (!originalMemberNames.has(name)) {
+        context.mismatches.report(original, `adds requirement for subclasses to implement '${name}'.`);
+      }
+    }
+}
+
+function describeTypeMatchingFailure(origType: reflect.TypeReference, updatedType: reflect.TypeReference, analysis: FailedAnalysis) {
+  if (origType.toString() !== updatedType.toString()) {
+    return `${updatedType} (formerly ${origType}): ${analysis.reasons.join(', ')}`;
+  } else {
+    return `${updatedType}: ${analysis.reasons.join(', ')}`;
+  }
+}
+
+function compareMethod(origClass: reflect.Type, original: reflect.Method, updated: reflect.Method, context: ComparisonContext) {
+  if (original.static !== updated.static) {
+    // tslint:disable-next-line:max-line-length
+    context.mismatches.report(origClass, `method ${original.name} was ${original.static ? 'static' : 'not static'}, is now ${updated.static ? 'static' : 'not static'}.`);
```

Comment:
Sure, but at least in Java it is not allowed to have a static an nonstatic member with the same name, so it should not be allowed in JSII. Hence, we are not losing fidelity by comparing interfaces this way.

Whether the `jsii` compiler checks that limitation is another matter. It should.

Reply:
Oh. I didn't know Java disallowed that (I guess I never tried, since it's relatively poor API design 🤷🏻‍♂️). I don't believe `jsii` checks for that... And yes, it should. Care to file an issue?

Reply:
Done.